docs/combinatorics.rst

```diff
@@ -1,6 +1,202 @@
 .. _sec_combinatorics:
 
-=====================================
-Ranking and Unranking Tree Topologies
-=====================================
-TODO
+=============
+Combinatorics
+=============
+This module implements a combinatorial approach to counting topologies
+of leaf-labelled trees. It provides methods for enumerating
+all possible trees with a given number of labelled leaves, as well as
+converting back and forth between a tree and its position, or rank, in
+the enumeration of all possible topologies.
+
++++++++++++++++++++++++++++++++++
+Ranking/Unranking Topologies
++++++++++++++++++++++++++++++++++
+The enumeration of leaf-labelled trees discussed below allows for an
+efficient method to identify such trees.
+By using the position, or rank, of a tree in the
+enumeration, we obtain a reversible hash for leaf-labelled trees. We
+can compute this rank without generating intermediate trees by relying
+on the algorithm used to generate them and formulas for knowing how
+many trees of a particular class there are. Similarly, given a rank,
+we can construct, or unrank, the corresponding tree. Below are definitions
+relating to tree ranks used in this process.
+
+***********
+Definitions
+***********
+
+shape rank
+    For a tree :math:`T` with :math:`n` leaves, the shape rank is the position
+    of :math:`T`'s shape in the enumeration of all tree shapes with
+    :math:`n` leaves.
+
+label rank
+    For a tree :math:`T`, the label rank is the position of the labelling of
+    :math:`T`'s leaves in the enumeration of all unique ways the leaves of
+    :math:`T` could be labelled.
+
+rank
+    The rank of a leaf-labelled tree is considered to be the tuple of its
+    ``shape rank`` and ``label rank``.
+
+unrank
+    The action of reconstructing a leaf-labelled tree from its ``rank``.
+
+++++++++++++++++++++++
+Enumerating Topologies
+++++++++++++++++++++++
+
+Below we outline the approach for enumerating all unique topologies
+of leaf-labelled trees with :math:`n` leaves. This takes part in two
+steps. First, we must consider how to generate unique tree shapes.
+Then, for each shape we must consider how to uniquely assign labels
+to the leaves. Note that this enumeration is not limited to binary trees.
+
+***********************
+Enumerating Tree Shapes
+***********************
+Let :math:`S_n` denote the set of all unlabelled trees, or tree shapes
+of :math:`n` leaves.
+We can see that :math:`S_1` has one element: a single leaf.
+
+A tree with :math:`n > 1` leaves can be obtained by joining at least two trees whose number
+of leaves sum to :math:`n`. This maps very closely to the concept of integer partitions.
+We first compute all integer partitions* of `n`, and then for each partition :math:`P`,
```

Comment:
shouldn't you say "all nondecreasing integer partitions", and explain what "nondecreasing" means?